### PR TITLE
Refactor server

### PR DESCRIPTION
**Pull request by lucasvreis**
Simplifies the websocket loop logic from "wait for two messages, first
one tells the current open route (to be watched), second one tells the
route to be sent" to "on every message, switch to its route and remember
that it is being watched".

Removes routeFromPathInfo (that would always strip the first character
from the websocket message) and instead assumes the js shim strips the
base path completely.

This is in preparation to the server customization feature.

Also simplify parts of the runServerWithWebSocketHotReload function.

### CODE REVIEW

**Comment by srid** [comment]
The diff of the original PR was harder to review. I pushed a few commits to make that easier.

File: ema/src/Ema/Server.hs [hunk 1]
```diff
@@ -95,56 +93,44 @@ runServerWithWebSocketHotReload host mport model = do
...
                     Right (Just r) -> do
-                      renderCatchingErrors s r >>= \case
+                      renderCatchingErrors m r >>= \case
                         AssetGenerated Html html ->
                           liftIO $ WS.sendTextData conn $ html <> toLazy wsClientHtml
                         -- HACK: We expect the websocket client should check for REDIRECT prefix.
                         -- Not bothering with JSON response to avoid having to JSON parse every HTML dump.
                         AssetStatic _staticPath ->
-                          liftIO $ WS.sendTextData conn $ "REDIRECT " <> toText (review (fromPrism_ $ enc s) r)
+                          liftIO $ WS.sendTextData conn $ "REDIRECT " <> toText (review (fromPrism_ $ routePrism m) r)
```
Comment (srid) [inline ema/src/Ema/Server.hs:112]: @lucasvreis What is the reason for adjacent changes like this?

- Replace `s` with `m`
- Replace `enc` with `routePrism`

I am just trying to narrow down to the "Simplifies the websocket loop logic" part, which is what I mostly care about.

Comment (srid) [reply ema/src/Ema/Server.hs:112]: I think a PR that does just the "Simplifies the websocket loop logic" part, but without the refactoring, would be most helpful.

Comment (srid) [reply ema/src/Ema/Server.hs:112]: I assume that the main impact of this PR is that it fixes the double-fetch, which alone would be cause for making another release of Ema.

Comment (lucasvreis) [reply ema/src/Ema/Server.hs:112]: > What is the reason for adjacent changes like this?

zero particular reasons. I just did git amend and undid some of the changes of the second commit until it looked reasonable. I was not caring about minimizing diffs or anything, so those lines were already shadowed by the indentation change.

Comment (lucasvreis) [reply ema/src/Ema/Server.hs:112]: > I think a PR that does just the "Simplifies the websocket loop logic from ... to ..." part, but without the refactoring, would be most helpful.

Understandable, I will reopen with more things split.

**Pull request closed without merging**